src/main.cpp

```diff
@@ -5668,6 +5668,18 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             RelayTransaction(tx);
             vWorkQueue.push_back(inv.hash);
 
+            // Add items to the workqueue if there are any.  We only add a few at a time
+            // so as not to keep the cs_main lock too long.  But we must add at least two because
+            // the workqueue is comprised of all the tx hashes for the last two blocks.
+            for (int i = 1; i <= 20; i++) {
+                uint256 pvHash = PV.NextWorkQueueItem();
+                if (!pvHash.IsNull())
+                    vWorkQueue.push_back(pvHash);
+                else
+                    break;
+            }
+
+
```

Comment:
If I understand this correctly, this code injects up to 20 tx that need to be reevaluated from the prior 2 blocks into the flow of processing of an incoming tx.  I am worried that this strategy may be confusing to readers and create bugs.  We are supposed to be processing the incoming transaction, but we go off and do an unrelated thing adding complexity.  So I'd like to see the orphan transaction processing logic factored out of main and then called by both the incoming tx and the PV work queue needs-reevaluation tx separately.

As a concrete example of my concerns:
1. we only process the tx in the PV workqueue if another transaction comes in... this is a small issue because lots of tx are always coming in on mainnet.  But what about our test networks?  Should we really have to generate tx to trigger 20 tx reevaluations?
2. if the "fromPeer" is "misbehaving" (line 5710) we skip the processing loop.  But this fromPeer did not necessarily send the transactions from the PV workqueue so should not affect their processing.
3. The inverse of 2: (line 5724) if a transaction from the PV workqueue is bad, we punish fromPeer, who did not necessarily send us this transaction.